src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -0,0 +1,231 @@
+
+    private static final String JAVA_PLUGIN = "org.gradle.java";
+
+    private static final String JAVA_LIBRARY_PLUGIN = "org.gradle.java-library";
+
+    private static final String APPLICATION_PLUGIN = "application";
+
+    private static final String EXTENSION_NAME = "architecture";
+
+    private static final String LIBS_PLACEHOLDER = "APP_HOME_LIBS_PLACEHOLDER";
+
+    @Override
+    public void apply(Project project) {
+        LOGGER.debug("Applying JigsawPlugin to " + project.getName());
+        if (!project.getPlugins().hasPlugin(JAVA_PLUGIN) && !project.getPlugins().hasPlugin(JAVA_LIBRARY_PLUGIN)) {
+            project.getPluginManager().apply(JavaPlugin.class);
+        }
+        project.getExtensions().create(EXTENSION_NAME, Architecture.class);
+
+        configureJavaTasks(project);
+    }
+
+    private void configureJavaTasks(final Project project) {
+        project.afterEvaluate(new Action<Project>() {
+            @Override
+            public void execute(final Project project) {
+                configureCompileJavaTask(project);
+                configureCompileTestJavaTask(project);
+                configureJarTask(project);
+                configureTestTask(project);
+                project.getPluginManager().withPlugin(APPLICATION_PLUGIN, new Action<AppliedPlugin>() {
+                    @Override
+                    public void execute(AppliedPlugin appliedPlugin) {
+                        configureRunTask(project);
+                        configureStartScriptsTask(project);
+                    }
+                });
+            }
+        });
+    }
+
+    private void configureCompileJavaTask(final Project project) {
+        final JavaCompile compileJava = (JavaCompile) project.getTasks().findByName(JavaPlugin.COMPILE_JAVA_TASK_NAME);
+        final Architecture module = (Architecture) project.getExtensions().getByName(EXTENSION_NAME);
+        final String moduleDir =
+                compileJava.getDestinationDir().toString() + System.getProperty("path.separator") + module.getModuleName();
+        compileJava.getInputs().property("moduleName", module.getModuleName());
+        compileJava.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                List<String> args = new ArrayList<>();
+                args.add("--module-path");
+                args.add(compileJava.getClasspath().getAsPath());
+                args.add("-d");
+                args.add(moduleDir);
+                compileJava.getOptions().setCompilerArgs(args);
+                compileJava.setClasspath(project.files());
+            }
+        });
+        compileJava.doLast(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                project.copy(new Action<CopySpec>() {
+                    @Override
+                    public void execute(CopySpec copySpec) {
+                        copySpec.from(moduleDir);
```

Comment:
Hmmm. We should probably have a chat about all of this. I think our understandings are not matching up. For example, I don't agree that the designers of the JDK think that multiple modules in the same directory on the `--module-path` is an edge case. They may not often be expecting them to be exploded in a common scenario. But, it is very much a common pattern in their getting started guide to build modules into some `mods` directory with a subdirectory for each module's classes and then to run with `--module-path mods` at runtime.

Also, there really isn't any reason that Gradle couldn't be rejiggered to build multiple java modules from a single Gradle project. We *might* not want to allow that, but we might. Lastly, I don't *think* that you actually are allowed to spread a single module over multiple directories on the `--module-path`. You can have some of the classes (including the `module-info.class`) on the `--module-path` and then patch in classes from some other directories using `--patch-module`. But, I think there is a requirement that (outside of `--patch-module`) the complete classes of the module be rooted in the same directory in the filesystem.  I need to test that assertion though. It would be interesting to learn that it is wrong.

Maybe you're thinking of the `--module-source-path` which definitely *can* have multiple directories from around the filesystem contributing to a single module.  But, that's at compilation time.

I am very interested in learning how I might be able to use variants to "mark up the compileClasspath to consume the modularized variant instead of the non-modularized classes dir." That sounds like powerful magic. Where can I see an example of how variants work? Has it made it into the user guide yet, or is it just in the code?